File: app-service-comments/src/main/java/com/github/arhor/aws/graphql/federation/comments/api/graphql/dataloader/AbstractMappedBatchLoader.java
package com.github.arhor.aws.graphql.federation.comments.api.graphql.dataloader;

import lombok.RequiredArgsConstructor;
import org.dataloader.MappedBatchLoader;

import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.function.Function;

import static java.util.Collections.emptyMap;

@RequiredArgsConstructor
public abstract class AbstractMappedBatchLoader<K, V> implements MappedBatchLoader<K, V> {

    private final Executor executor;
    private final Function<Set<K>, Map<K, V>> loaderFn;

    @Override
    public CompletableFuture<Map<K, V>> load(
        final Set<K> keys
    ) {
        return keys.isEmpty()
            ? CompletableFuture.completedFuture(emptyMap())
            : CompletableFuture.supplyAsync(() -> loaderFn.apply(keys), executor);
    }
}

File: app-service-posts/src/main/kotlin/com/github/arhor/aws/graphql/federation/posts/infrastructure/graphql/dataloader/AbstractMappedBatchLoader.kt
package com.github.arhor.aws.graphql.federation.posts.infrastructure.graphql.dataloader

import org.dataloader.MappedBatchLoader
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executor

abstract class AbstractMappedBatchLoader<K, V>(
    private val executor: Executor,
    private val loaderFn: (Set<K>) -> Map<K, V>,
) : MappedBatchLoader<K, V> {

    @Override
    override fun load(keys: Set<K>): CompletableFuture<Map<K, V>> =
        if (keys.isEmpty()) {
            CompletableFuture.completedFuture(emptyMap())
        } else {
            CompletableFuture.supplyAsync({ loaderFn.invoke(keys) }, executor)
        }
}
